compiler/src/dmd/semantic3.d

```diff
@@ -342,6 +342,8 @@ private extern(C++) final class Semantic3Visitor : Visitor
             sc2.aligndecl = null;
             if (funcdecl.ident != Id.require && funcdecl.ident != Id.ensure)
                 sc2.flags = sc.flags & ~SCOPE.contract;
+            if (sc2.flags & SCOPE.compile)
+                funcdecl.skipCodegen = true;
             sc2.flags &= ~SCOPE.compile;
```

Comment:
Currently dmd has multiple ways of checking for speculative context, each with its subtle differences:
- sc.minst and sc.tinst are both null.
- sc.flags & SCOPE.compile
- FuncDeclaration.skipCodegen
- FuncDeclaration.isCompileTimeOnly

Here, SCOPE.compile is randomly cleared because it is for attribute inference in a different place.
The entire situation is crazy...

I'm looking into simplifying the mess, but until that point, I'm using `skipCodegen`  to signal typeid that we are in a speculative context. I tried all of the above, but only `skipCodegen` works.